cmd/kubelet/app/plugins_providers.go

```diff
@@ -28,12 +32,25 @@ import (
 	"k8s.io/kubernetes/pkg/volume/vsphere_volume"
 )
 
+type probeFn func() []volume.VolumePlugin
+
+func appendPlugin(plugins []volume.VolumePlugin, key featuregate.Feature, fn probeFn) []volume.VolumePlugin {
+	// Skip appending the in-tree plugin to the list of plugins to be probed/initialized
+	// if the CSIMigrationInTreeOff feature flag and plugin specific CSI migration
+	// feature flag is enabled
+	if feature.DefaultFeatureGate.Enabled(features.CSIMigrationInTreeOff) && feature.DefaultFeatureGate.Enabled(key) {
```

Comment:
it's hard to tell here if the feature gate is initialized at the point these are called. would it be feasible to pass in the feature-gate-holder that should be checked rather than adding dependencies on k8s.io/apiserver to more places in the kubelet?

Reply:
@liggitt Do you have a pointer to usage of a feature-gate-holder elsewhere in the kubelet code? In the above code, I check for the kubelet feature flag (specified in the kubelet CLI) to be enabled. It seems based on the kubelet [run](https://github.com/kubernetes/kubernetes/blob/ed3cc6afea6fa3d9f8e2a1544daaa12f87d2b65c/cmd/kubelet/app/server.go#L472) function, the feature gate [initialization](https://github.com/kubernetes/kubernetes/blob/ed3cc6afea6fa3d9f8e2a1544daaa12f87d2b65c/cmd/kubelet/app/server.go#L474) for the kubelet happens before [RunKubelet](https://github.com/kubernetes/kubernetes/blob/ed3cc6afea6fa3d9f8e2a1544daaa12f87d2b65c/cmd/kubelet/app/server.go#L755) is invoked (which ultimately registers the storage plugins). Given that structure of the code, is it possible for the feature gate flags to not be initialized in the context of the kubelet before we reach plugin initialization?

Reply:
> Do you have a pointer to usage of a feature-gate-holder elsewhere in the kubelet code?

Not in the kubelet, but https://github.com/kubernetes/kubernetes/pull/84813 is an example of plumbing the gate holder to the consumer

> It seems based on the kubelet run function, the feature gate initialization for the kubelet happens before RunKubelet is invoked (which ultimately registers the storage plugins). Given that structure of the code, is it possible for the feature gate flags to not be initialized in the context of the kubelet before we reach plugin initialization?

Probably not, though some of the dynamic kubelet config load paths are twisty.